provider.go

```diff
@@ -854,13 +820,138 @@ func (p *provider) Call(ctx context.Context, req *rpc.CallRequest) (*rpc.CallRes
+) (CallResponse, error) {
+	r, err := comProvider.Call(ctx, req.rpc(internalrpc.MarshalProperties), h.host.EngineConn(), call)
+	if err != nil {
+		return CallResponse{}, err
+	}
+	return newCallResponse(r)
+}
+
 // CallRequest represents a requested resource method invocation.
 //
 // It corresponds to [rpc.CallRequest] on the wire.
 type CallRequest struct {
-	Tok     tokens.ModuleMember
-	Args    presource.PropertyMap
-	Context *pulumi.Context
+	// the function token to invoke.
+	Tok tokens.ModuleMember
+	// the arguments for the function invocation.
+	Args presource.PropertyMap
+	// a map from argument keys to the dependencies of the argument.
+	ArgDependencies map[presource.PropertyKey][]presource.URN
+	// the project name.
+	Project string
+	// the name of the stack being deployed into.
+	Stack string
+	// the configuration variables to apply before running.
+	Config map[pconfig.Key]string
+	// the configuration keys that have secret values.
+	ConfigSecretKeys []pconfig.Key
+	// true if we're only doing a dryrun (preview).
+	DryRun bool
+	// the degree of parallelism for resource operations (<=1 for serial).
+	Parallel int32
+	// the address for communicating back to the resource monitor.
+	MonitorEndpoint string
+	// the organization of the stack being deployed into.
+	Organization string
+}
+
+func newCallRequest(req *rpc.CallRequest, unmarshal rpcToProperty) (CallRequest, error) {
+	args, err := unmarshal(req.GetArgs())
+	if err != nil {
+		return CallRequest{}, fmt.Errorf("unable to convert args into a property map: %w", err)
+	}
+
+	r := CallRequest{
+		Tok:  tokens.ModuleMember(req.GetTok()),
+		Args: args,
+		ArgDependencies: func() map[presource.PropertyKey][]presource.URN {
+			r := make(map[presource.PropertyKey][]presource.URN, len(req.GetArgDependencies()))
+			for k, v := range req.GetArgDependencies() {
+				r[presource.PropertyKey(k)] = make([]presource.URN, len(v.GetUrns()))
+				for i, urn := range v.GetUrns() {
+					r[presource.PropertyKey(k)][i] = presource.URN(urn)
+				}
+			}
+			return r
+		}(),
+		Project: req.GetProject(),
+		Stack:   req.GetStack(),
+		Config: func() map[pconfig.Key]string {
+			m := make(map[pconfig.Key]string, len(req.GetConfig()))
+			for k, v := range req.GetConfig() {
+				m[pconfig.MustParseKey(k)] = v
+			}
+			return m
+		}(),
+		ConfigSecretKeys: func() []pconfig.Key {
+			keys := make([]pconfig.Key, len(req.GetConfigSecretKeys()))
+			for i, k := range req.GetConfigSecretKeys() {
+				keys[i] = pconfig.MustParseKey(k)
+			}
+			return keys
+		}(),
+		DryRun:          req.GetDryRun(),
+		Parallel:        req.GetParallel(),
+		MonitorEndpoint: req.GetMonitorEndpoint(),
+		Organization:    req.GetOrganization(),
+	}
+	return r, nil
+}
+
+func (c CallRequest) rpc(marshal propertyToRPC) *rpc.CallRequest {
+
+	fromUrns := func(urns []presource.URN) []string {
+		r := make([]string, len(urns))
+		for i, urn := range urns {
+			r[i] = string(urn)
+		}
+		return r
+	}
+
+	// Marshal the args.
```

Comment:
```suggestion
```
Not sure this is necessary? Same for similar comments about unmarshalling later, but happy either way

Reply:
We need 100% fidelity going in/out of the rpc representation, to support `middleware.rpc` which calls into a legacy provider.